topology/probes/bess/bess.go

```diff
@@ -0,0 +1,248 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bess
+
+import (
+	"context"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"github.com/skydive-project/skydive/filters"
+	"google.golang.org/grpc"
+	"strings"
+	"time"
+
+	"github.com/nimbess/nimbess-agent/pkg/proto/bess_pb"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/probe"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/probes"
+)
+
+const BessBackend = "BESS"
+
+type Probe struct {
+	graph.DefaultGraphListener
+	Ctx         probes.Context
+	bessClient  bess_pb.BESSControlClient
+	state       common.ServiceState
+	bessPort    int
+	bessHost    string
+	bessContext context.Context
+}
+
+func (p *Probe) walkAndRender(moduleName string, parentNode *graph.Node, edgeList *[]*graph.Edge) {
+	modRes, err := p.bessClient.GetModuleInfo(p.bessContext, &bess_pb.GetModuleInfoRequest{Name: moduleName})
+	if err != nil {
+		log.Errorf("Error getting module information for module: %s", moduleName)
+		return
+	}
+
+	nodeID := graph.GenID(moduleName)
+	node := p.Ctx.Graph.GetNode(nodeID)
+	if node == nil {
+		node, err = p.Ctx.Graph.NewNode(nodeID, graph.Metadata{"Name": moduleName,
+			"Type": modRes.GetMclass(), "Backend": BessBackend})
+		log.Infof("Creating node: %s", moduleName)
+		if err != nil {
+			log.Errorf("Error creating node: %s, %v", moduleName, err)
+		}
+	}
+	if parentNode != nil && !topology.HaveOwnershipLink(p.Ctx.Graph, parentNode, node) {
+		topology.AddOwnershipLink(p.Ctx.Graph, parentNode, node, nil)
+	}
+	for _, oGate := range modRes.GetOgates() {
+		p.walkAndRender(oGate.GetName(), parentNode, edgeList)
+		edgeID := graph.GenID(moduleName, string(oGate.Igate), string(oGate.Ogate), oGate.GetName())
+		if p.Ctx.Graph.GetEdge(edgeID) == nil {
+			p.Ctx.Graph.NewEdge(edgeID, node,
+				p.Ctx.Graph.GetNode(graph.GenID(oGate.GetName())),
+				graph.Metadata{"RelationType": "layer2", "Directed": true, "Backend": BessBackend})
+		}
+		*edgeList = append(*edgeList, p.Ctx.Graph.GetEdge(edgeID))
+	}
+
+}
+
+func (p *Probe) run() {
+	p.state.Store(common.RunningState)
+	fmt.Println("Starting BESS Probe...")
+	logging.GetLogger().Info("Starting BESS Probe")
+	conn, err := grpc.Dial(fmt.Sprintf("%s:%d", p.bessHost, p.bessPort), grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("Did not connect: %v", err)
```

Comment:
Usually probe don't panic or exit, they just log.Error and keep trying to reconnect. I would suggest to use a `NodeProbeWrapper` which can handle reconnection loop.
https://github.com/skydive-project/skydive/blob/master/topology/probes/wrapper.go#L125

You can have a look how the `docker` probe is using it. Basically you run function will be replaced by a `Do` function

nit: use `%s` for error

Reply:
OK I'll look into this, thanks.